book/en/src/internals/tasks.md

```diff
@@ -1,3 +1,397 @@
+        // the priority of this interrupt handler
+        const PRIORITY: u8 = 1;
+
+        let snapshot = basepri::read();
+
+        while let Some(ready) = RQ1.split().1.dequeue() {
+            match ready.task {
+                T1::bar => {
+                    // **NOTE** simplified implementation
+
+                    // used to track the dynamic priority
+                    let priority = Cell::new(PRIORITY);
+
+                    // call into user code
+                    bar(bar::Context::new(&priority));
+                }
+
+                T1::baz => {
+                    // we'll look at `baz` later
+                }
+            }
+        }
+
+        // BASEPRI invariant
+        basepri::write(snapshot);
+    }
+};
+```
+
+## Spawning a task
+
+The `spawn` API is exposed to the user as the methods of a `Spawn` struct.
+There's one `Spawn` struct per task.
+
+The `Spawn` code generated by the framework for the previous example looks like
+this:
+
+``` rust
+mod foo {
+    // ..
+
+    pub struct Context<'a> {
+        pub spawn: Spawn<'a>,
+        // ..
+    }
+
+    pub struct Spawn<'a> {
+        // tracks the dyanmic priority of the task
+        priority: &'a Cell<u8>,
+    }
+
+    impl<'a> Spawn<'a> {
+        // `unsafe` and hidden because we don't want the user to tamper with it
+        #[doc(hidden)]
+        pub unsafe fn priority(&self) -> &Cell<u8> {
+            self.priority
+        }
+    }
+}
+
+const APP: () = {
+    // ..
+
+    // Priority ceiling for the producer endpoint of the `RQ1`
+    const RQ1_CEILING: u8 = 2;
+
+    // used to track how many more `bar` messages can be enqueued
+    // `U2` is the capacity of the `bar` task; a max of two instances can be queued
+    // this queue is filled by the framework before `init` runs
+    static mut bar_FQ: Queue<(), U2> = Queue::new();
+
+    // Priority ceiling for the consumer endpoint of `bar_FQ`
+    const bar_FQ_CEILING: u8 = 2;
+
+    // a priority-based critical section
+    //
+    // this run the given closure `f` at a dynamic priority of at least
+    // `ceiling`
+    fn lock(priority: &Cell<u8>, ceiling: u8, f: impl FnOnce()) {
+        // ..
+    }
+
+    impl<'a> foo::Spawn<'a> {
+        /// Spawns the `bar` task
+        pub fn bar(&self) -> Result<(), ()> {
+            unsafe {
+                match lock(self.priority(), bar_FQ_CEILING, || {
+                    bar_FQ.split().1.dequeue()
+                }) {
+                    Some(()) => {
+                        lock(self.priority(), RQ1_CEILING, || {
+                            // put the taks in the ready queue
+                            RQ1.split().1.enqueue_unchecked(Ready {
+                                task: T1::bar,
+                                // ..
+                            })
+                        });
+
+                        // pend the interrupt that runs the task dispatcher
+                        rtfm::pend(Interrupt::UART0);
```

Comment:
I'm currently wondering if this needs a compiler fence to prevent compiler
optimizations from changing the intended semantics.

`rtfm::pend` is a volatile write but `unqueue_unchecked` is a non-volatile
operation so the compiler is free (to some extend) to reorder these two
operations. If we zoom in into these two statements:

``` rust
lock(self.priority(), RQ1_CEILING, || {
    // put the taks in the ready queue
    RQ1.split().1.enqueue_unchecked(Ready {
        task: T1::bar,
        // ..
    })
});

// pend the interrupt that runs the task dispatcher
rtfm::pend(Interrupt::UART0);
```

Assuming that the dynamic priority is equal to the priority ceiling of `RQ1` we
can omit the critical section thus we end up with:

``` rust
RQ1.split().1.enqueue_unchecked(Ready {
    task: T1::bar,
    // ..
})

rtfm::pend(Interrupt::UART0);
```

If we then inline all operations to the maximum we get:

``` rust
// `RQ1` has type `Queue<Ready, const N>` which is equal to
// `struct { tail: AtomicU8, head: AtomicU8, buffer: [Ready; N] }`

/* `RQ1.split().1.enqueue_unchecked(..)` */
let tail = RQ1.tail.load(Ordering::Relaxed);
let ready = Ready { task: T1::bar, /* .. */ };

RQ1.buffer.get_unchecked_mut((tail as usize) % N).write(ready);

atomic::compile_fence(Ordering::Release); // ▲

RQ1.tail.store(tail + 1, Ordering::Relaxed);

/* `rtfm::pend(Interrupt::UART0)` */
ptr::write_volatile(0xE000E200, 1 << 4);
```

The `Release` fence means "preceding reads and writes cannot be moved past
subsequent writes". This is required for memory safety: the `buffer.write`
should *not* be reordered to after the `tail.store`.

However, I think the compiler is allowed to swap the `tail.store` and
`write_volatile` operations. The consequence of that would be that `UART0`
handler could fire and run the task dispatcher which would then see the *old*
`tail` value and miss the just enqueued entry. This doesn't break memory safety
but it would cause the `spawn`-ed task to be missed and not run until some other
task is `spawn`-ed.

I think to be on the safe side we should add another `Release` fence right before
the `write_volatile` (`rtfm::pend`). That is:

``` rust
// ..

RQ1.buffer.get_unchecked_mut((tail as usize) % N).write(ready);

atomic::compile_fence(Ordering::Release); // ▲

RQ1.tail.store(tail + 1, Ordering::Relaxed);

// NEW!
atomic::compile_fence(Ordering::Release); // ▲

/* `rtfm::pend(Interrupt::UART0)` */
ptr::write_volatile(0xE000E200, 1 << 4);
```

Now the compiler can't swap `tail.store` and `write_volatile`. In the generated
code this would be reflected as:

``` rust
lock(self.priority(), RQ1_CEILING, || {
    // put the taks in the ready queue
    RQ1.split().1.enqueue_unchecked(Ready {
        task: T1::bar,
        // ..
    })
});

// NEW! complete the `enqueue` operation before pending the interrupt
atomic::compile_fence(Ordering::Release); // ▲

// pend the interrupt that runs the task dispatcher
rtfm::pend(Interrupt::UART0);
```

I think this should not negatively impact optimizations as there's already a
compiler fence in `enqueue_unchecked`.

Reply:
Taking a step back:

So what you are touching upon here boils down to what model of computation / semantics we want.

Resources as critical sections.
From a modelling perspective, one can see Resources as specifications of named critical sections, i.e., a way to ensure ordering and mutual exclusion.

From a pure modelling perspective pend is an "async" in RTFM. "async" semantics is that they will execute independently to the sender, inheriting the permissible execution window (base and deadline). "asyncs" can have payload in RTFM, the semantics follow that of sequential ordering data dependencies (i.e., parameters should be computed first). Side effects of synchronous operations before the "async" should be observable to the "pended" task.

Under that model even if the actual BASEPRI does not need to be manipulated, we should have barriers around the lock (to reflect the model).

This have not been formalized but builds on the Timber ideas. 

For this case a single barrier might be sufficient to uphold the model above. (I recall from the original paper RTFM Step 1, that we enforced  barriers at both entry and exit to be on the safe side).

Side note:
In Rust-RTFM we don't yet have a formalized semantics. We can define it the way we want (the only real requirement is that it does not break SRP, and does not break the Rust ownership semantics). What we deal with here is outside of these requirements, but very important to lay the foundation for reasoning about programs in Rust-RTFM, and about giving the user fine grained control.

I think the idea/notion of Resources as named critical sections is a very nice abstraction, allowing you to define logical resources (by maker typed static muts), and use those to provide fine grained ordering and concurrency control. (Of course it also covers the standard use of resources to safely share mutable data).

/Per